### PR TITLE
Notification with actions

### PR DESCRIPTION
**Pull request by thearaks**
Hi all!
Here I submit a Notification Actions implementation.
This adds the "Open app" and "Stop" actions and improves the overall notification style and attributes on latest Android versions.

There's also a small change to MainActivity to ensure that its GUI reflects the service state (started/stopped).

Removed unused string (notification text).

### CODE REVIEW

**Comment by DJCrapsody** [comment]
Some weird behaviours I have encountered and steps to reproduce them:
* IV button goes missing sometimes
 1. Start GoIV
 2. Pokemon Info (IV button appears)
 3. Stop GoIV from Notification
 4. Start GoIV
 5. Pokemon Info (no IV button)
 6. Stop GoIV from Notification
 7. Start GoIV
 8. Pokemon Info (IV button appears)

* Stop button does not change back to Start
 1. Open GoIV using Recent Apps or from Notification
 2. Stop GoIV from Notification (Button still stays as 'Stop' with 'Stop button onClick' event (click and it turns to Start)
 3. Leave GoIV (go to another app from Recent Apps or go to Home Screen)
 4. Return to GoIV (Button changes to 'Start' with 'Start button onClick' event (click and it starts Pokemon GO)

File: app/src/main/java/com/kamron/pogoiv/MainActivity.java [hunk 1]
```diff
@@ -308,7 +309,14 @@ public void onClick(DialogInterface dialog, int which) {
...
     protected void onResume() {
         super.onResume();
+
         settings = GoIVSettings.getInstance(MainActivity.this);
+
+        if (isPokeFlyRunning()) {
+            launchButton.setText(R.string.main_stop);
+        } else {
+            launchButton.setText(R.string.main_start);
+        }
```
Comment (Blaisorblade) [inline app/src/main/java/com/kamron/pogoiv/MainActivity.java]: @DJCrapsody's testing suggests calling `launchButton.setText` here is not enough, it should also be called when starting/stopping the service.

File: app/src/main/java/com/kamron/pogoiv/MainActivity.java [hunk 2]
```diff
@@ -84,9 +85,10 @@
 
     private boolean readyForNewScreenshot = true;
 
-    private boolean pokeFlyRunning = false;
```
Comment (Blaisorblade) [inline app/src/main/java/com/kamron/pogoiv/MainActivity.java]: 👍

File: app/src/main/java/com/kamron/pogoiv/Pokefly.java [hunk 3]
```diff
@@ -436,11 +443,12 @@ private void hideInfoLayoutArcPointer() {
...
     @Override
     public void onDestroy() {
-        if (!batterySaver) {
+        super.onDestroy();
+
+        if (!batterySaver && timer != null) {
             timer.cancel();
         }
 
-        super.onDestroy();
```
Comment (Blaisorblade) [inline app/src/main/java/com/kamron/pogoiv/Pokefly.java]: IIRC super.onDestroy had to be placed after `timer.cancel()`—and ideally it'd go at the end. I'd leave out unrelated changes.

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -16,7 +16,7 @@
     <string name="main_not_numeric">%1$s is not a number.</string>
     <string name="main_invalide_trainerlvl">Invalid Trainer Level!</string>
     <string name="notification_title">GoIV Running - Level %1$s</string>
-    <string name="notification_text">Tap to open</string>
+    <string name="notification_open_app">Open app</string>
```
Comment (Blaisorblade) [inline app/src/main/res/values/strings.xml]: Is there a behavior change that requires changing (and retranslating) the string?

File: app/src/main/java/com/kamron/pogoiv/Pokefly.java [hunk 5]
```diff
@@ -456,21 +464,35 @@ public void onDestroy() {
      * makeNotification
      * Creates the GoIV notification
      */
-    private void makeNotification(Context context) {
-        Intent intent = new Intent(context, MainActivity.class);
+    private void makeNotification() {
+        Intent openAppIntent = new Intent(this, MainActivity.class);
+
+        PendingIntent openAppPendingIntent = PendingIntent.getActivity(this,
+                0, openAppIntent, PendingIntent.FLAG_UPDATE_CURRENT);
```
Comment (Blaisorblade) [inline app/src/main/java/com/kamron/pogoiv/Pokefly.java]: Why 0? Isn't this supposed to match `NOTIFICATION_REQ_CODE` below?

File: app/src/main/java/com/kamron/pogoiv/MainActivity.java (hunk 1 above)
Comment (Blaisorblade) [inline app/src/main/java/com/kamron/pogoiv/MainActivity.java]: Also, the existing logic for starting the service only works correctly if the label is correct (which seems... weird to me). @DJCrapsody in your first scenario, is it possible the label hadn't been updated? Because the logic to start the service from the button is unchanged.

**Comment by Blaisorblade** [comment]
Thanks for working on this!

I have to give it a try (wrong computer); meanwhile I've taken a look at the code and added some initial comments. Feel free to correct me if I'm wrong, I am aware I'm not an Android expert.

**Comment by thearaks** [comment]
Thanks @Blaisorblade and @DJCrapsody,
I'll check everything as soon as I'll have a spare hour at home!

I think I might use the LocalBroadcastManager to keep the GUI in sync as the service changes state.
Also, the start/stop controls logic should be brought out of the launch button OnClickListener. I'll study a cleaner solution.

About super.onDestroy() you're right: it should be at the end, I don't even know why I touched that line...

Thank you again!

**Comment by Blaisorblade** [comment]
Thanks for your positive response! Looking forward to your new code (when you have time of course).

> I think I might use the LocalBroadcastManager to keep the GUI in sync as the service changes state.

`LocalBroadcastManager` can be a source of pain—which is why I thought we needed to refactor the app to allow for this feature. Maybe there's a way out though.

If you do that, you might want to make sure the code is robust to duplicated events (apparently broadcasts can get duplicated on some devices). Ideally, calling the handler the second time would be harmless (say because it repeats the same assignments)—technically, the handler should be idempotent. The code you added in `onResume` satisfies that condition.

FWIW: Currently we use a hack which causes trouble—an intent changes state so that a repeated intent is rejected. But those flags get out of sync and reject valid intents (causing all the "IV button disappeared" bugs). And fixing the "flags get out of sync" would involve ensuring hundreds of lines of code stop failing. As you'd expect, that's not easy 😉

**Comment by Blaisorblade** [comment]
I'm closing this for now as the bug is tracked by #268 (and I'm trying to process PRs/issues), but I'm looking forward to an updated PR!

**Comment by thearaks** [comment]
Hi @Blaisorblade,
I have a new version of this PR but I can't fix the Start / Stop from Notification bug found by @DJCrapsody. 
When Pokefly stops itself from the Notification Intent instead of MainActivity then it can't start up correctly anymore.
I'm investigating this strange behavior right now...
